src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs

```diff
@@ -237,18 +211,31 @@ private void ScanInterpolatedStringLiteral(bool isVerbatim, ref TokenInfo info)
             // /**/ comments, ' characters quotes, () parens
             // [] brackets, and "" strings, including interpolated holes in the latter.
 
-            SyntaxDiagnosticInfo error = null;
-            bool closeQuoteMissing;
-            ScanInterpolatedStringLiteralTop(null, isVerbatim, ref info, ref error, out closeQuoteMissing);
+            SyntaxDiagnosticInfo? error = null;
+            ScanInterpolatedStringLiteralTop(interpolations: null, isVerbatim, ref info, ref error, out _);
             this.AddError(error);
         }
 
-        internal void ScanInterpolatedStringLiteralTop(ArrayBuilder<Interpolation> interpolations, bool isVerbatim, ref TokenInfo info, ref SyntaxDiagnosticInfo error, out bool closeQuoteMissing)
+        internal void ScanInterpolatedStringLiteralTop(
+            ArrayBuilder<Interpolation>? interpolations, bool isVerbatim, ref TokenInfo info, ref SyntaxDiagnosticInfo? error, out bool closeQuoteMissing)
         {
             var subScanner = new InterpolatedStringScanner(this, isVerbatim);
             subScanner.ScanInterpolatedStringLiteralTop(interpolations, ref info, out closeQuoteMissing);
-            error = subScanner.error;
-            info.Text = TextWindow.GetText(false);
+
+            // if there are two syntax errors reported, then notify the user of the one earliest in the file.
```

Comment:
Have you considered only holding onto only one error (like before), but adding a flag to signify we've encountered an unrecoverable error?
If we did that, I suspect the earliest error would naturally win (as it would be produced first).

Reply:
yeah, i actually started with that pattern, but i did find it a little more confusing in terms of how every place that updated errors had to manage checking that and processing accordingly.  i actually thought this was really clean in terms of what was going on, with only the additional wrinkle of checking the starting position.

but i may not toally be getting your idea here.  i'd be happy to collaborate on this :)

Reply:
Just to clarify my idea (maybe it doesn't work that well, not sure):
1. `InterpolatedStringScanner` has `public SyntaxDiagnosticInfo error;` and `public bool encounteredUnrecoverableError = false;`
2. All places that do `UnrecoverableError = ...` or `UnrecoverableError ??= ...` use a helper method (`SetUnrecoverableError`) that sets `error` if it's the first one and `encounteredUnrecoverableError = true`.
3. The one place that does `RecoverableError ??= ...` can just do `error ??= ...`
4. `RecoveringFromRunawayLexing` checks `encounteredUnrecoverableError`
5. We should be able to remove the handling in `ScanInterpolatedStringLiteralTop` for picking which error to report

Reply:
ok.  i'll try that!

Reply:
Ok.  Done in https://github.com/dotnet/roslyn/pull/54875/commits/59e158d6cbd4b92917e4f433d4c82b36e798dbef.  Pro: it works.  Con: it technically means we may allocate more in error cases as we now always make the error prior to checking if we already have an error.

I could update this to pass in arguments, but there are some params cases, and we'll have to allocate.  

THat said, it's likely not a big deal as this is only for error conditions and it's unlikely that this would matter in practice.  Up to you if we should keep this or rollback :)

Reply:
I think it's a positive change. I'm not worried about this allocation, but passing in arguments would seem sensible if it helps. Worst case it's possible to add a null check at callsite.
Thanks!